### DISCUSSION TITLE
Extracting Slice Data

### DISCUSSION THREAD

**Discussion by rbielaw** [AMReX-Codes/amrex#3144] [Q&A] [answered]
Hey, I've been working on trying to add in run-time data extraction to our solver and was wondering if there was a way to extract a slice during run-time and write that slice as a PLT file?  I was trying to use get_slice_data function for this but was having some trouble getting it to work. Are there a set of functions to do this already or do you have a recommendation for the best way to do this?

**Comment by WeiqunZhang** [comment, marked as answer]
> I was trying to use get_slice_data function for this but was having some trouble getting it to work.

More detail?

Is your data on a single level or multiple levels?  If it's on multiple levels, do you want the slice to be at the finest resolution or the coarsest level or a multiple levels or something else?  How do you want to select the location of the slice?  Integer index or floating point number location?  If it's the latter, do you want just the nearest point or interpolation?

**Reply by rbielaw** [reply]
For get_slice_data my understanding is it returns the slice on a single level, and I was unable to figure out how to assemble this data from multiple levels so I could write it into a file. 

The data is on multiple levels, and I want the slice to be on multiple levels. The locations of the slice would be set by a floating point number location; the nearest neighbor is accurate enough for our applications. The reason for doing this is the full solution is too large to write frequently (~500GB) and so writing slices allows for higher frequency output in key locations.

**Reply by WeiqunZhang** [reply]
There is a function in AMReX_PlotFileUtil.H
```
    void WriteMultiLevelPlotfile (const std::string &plotfilename,
                                  int nlevels,
                                  const Vector<const MultiFab*> &mf,
                                  const Vector<std::string> &varnames,
                                  const Vector<Geometry> &geom,
                                  Real time,
                                  const Vector<int> &level_steps,
                                  const Vector<IntVect> &ref_ratio,
                                  const std::string &versionName = "HyperCLaw-V1.1",
                                  const std::string &levelPrefix = "Level_",
                                  const std::string &mfPrefix = "Cell",
                                  const Vector<std::string>& extra_dirs = Vector<std::string>());
```
You should be able to use it to write a plotfile of the multi-level slice data.  Note that you need to create some new 2D Geometry objects.

**Reply by rbielaw** [reply]
I'm still having some trouble assembling the geometry and FABS correctly. The following code is able to write the slice with the data as a PLT file, but all AMR levels appear at the same time when plotting in Paraview as opposed to only showing the highest AMR level at a location. Am I missing some step regarding the connectivity of the geometry, or do you have any guidance on why this might be occurring? Thanks for all the help so far.  

     int dir = 2; 
     amrex::Real dir_coord(0.0);
     Vector<Geometry> geomarr; 
     Vector<std::unique_ptr<MultiFab>> vec_slice;

     for(int sliceLevel = sliceStartLevel; sliceLevel <= sliceEndLevel; sliceLevel++){  
          
          HMM& lev_slice = getLevel(sliceLevel); 
          const Geometry& geom_full = lev_slice.geom;
          MultiFab&  S_full   = lev_slice.get_new_data(State_Type);
          //Sliced data
          std::unique_ptr<MultiFab> slice_data = get_slice_data(dir, dir_coord, S_full, geom_full, 0, NUM_STATE, false);   

          //Build Geometry for the slice
          RealBox real_slice = geom_full.ProbDomain();
          real_slice.setLo(dir, dir_coord);
          real_slice.setHi(dir, dir_coord);
          IntVect slice_lo, slice_hi;
          AMREX_D_TERM(slice_lo[0]=static_cast<int>(std::floor((real_slice.lo(0) - geom_full.ProbLo(0))/geom_full.CellSize(0)));,
                       slice_lo[1]=static_cast<int>(std::floor((real_slice.lo(1) - geom_full.ProbLo(1))/geom_full.CellSize(1)));,
                       slice_lo[2]=static_cast<int>(std::floor((real_slice.lo(2) - geom_full.ProbLo(2))/geom_full.CellSize(2))););
          AMREX_D_TERM(slice_hi[0]=static_cast<int>(std::floor((real_slice.hi(0) - geom_full.ProbLo(0))/geom_full.CellSize(0)));,
                       slice_hi[1]=static_cast<int>(std::floor((real_slice.hi(1) - geom_full.ProbLo(1))/geom_full.CellSize(1)));,
                       slice_hi[2]=static_cast<int>(std::floor((real_slice.hi(2) - geom_full.ProbLo(2))/geom_full.CellSize(2))););
          Box slice_box = Box(slice_lo, slice_hi) & geom_full.Domain();  
          amrex::Box domain_of_slice(slice_lo, slice_hi);
          amrex::Geometry geom_of_slice(domain_of_slice, &real_slice); 
          //Fill Geometry and MultiFab Vectors
          geomarr.emplace_back(geom_of_slice);        
          vec_slice.emplace_back(std::move(slice_data));
     }

     Vector<int> lsarr {sliceStartLevel, sliceEndLevel };
     amrex::WriteMultiLevelPlotfile (sliceFileName,
                                     nlevels+1,
                                     amrex::GetVecOfConstPtrs(vec_slice),
                                     VarNames,
                                     geomarr,
                                     time,
                                     lsarr,
                                     rrarr,
                                     "HyperCLaw-V1.1",
                                     "Level_",
                                     "Cell");

**Reply by WeiqunZhang** [reply]
The issue might be how `Box domain_of_slice` is created.  Could you try something like
```
Box slice_box = geom_full.Domain();
slice_box.setRange(dir, slice_data->boxArray()[0].smallEnd(dir));
```

What's `nlevels`? Why is `nlevels+1`?  Why not just `vec_slice.size()`?

Is sliceStartLevel == 0?  If not, the coarse level does not cover the whole domain.  I am not sure if that could be an issue.

**Reply by rbielaw** [reply]
SliceStartLevel is equal to zero. Thank you for all the help; this change allowed the slice to retain the AMR connectivity.

**Comment by drummerdoc** [comment]
@esclapez has something that will do exactly this.  It creates a 2D plotfile from the 3D data.
